_sips/sips/2017-09-20-opaque-types.md

```diff
@@ -736,6 +753,207 @@ This example is somewhat similar to `Logarithm` above. Other
+    def binarySearch(ia: IArray[Long], elem: Long): Int =
+      Arrays.binarySearch(ia, elem)
+  }
+
+  // same as IArray.binarySearch but implemented by-hand.
+  //
+  // given a sorted IArray, returns index of `elem`,
+  // or a negative value if not found.
+  def binaryIndexOf(ia: IArray[Long], elem: Long): Int = {
+    var lower: Int = 0
+    var upper: Int = IArray.size(ia)
+    while (lower <= upper) {
+      val middle = (lower + upper) >>> 1
+      val n = IArray.get(ia, middle)
+
+      if (n == elem) return middle
+      else if (n < elem) first = middle + 1
+      else last = middle - 1
+    }
+    -lower - 1
+  }
+}
+```
+
+This example is a bit different from others: there's no attempt to
+enrich the `IArray` type with syntactic conveniences. Rather, the goal
+is to show that traditional, "low-level" code with `Array`, `Int`,
+etc. can be written with opaque types without sacrificing any
+performance.
+
+Our other examples enrich existing data types with new
+functionality. This example serves to constrain the operations used
+with a type (but without introducing any overhead/indirection, which a
+traditional wrapper would).
+
+## Differences with value classes
+
+Most of the above examples can also be implemented using value
+classes. This section will highlight some differences between these
+hypothetical encodings.
+
+### Used as a type parameter
+
+In many cases an author would introduce opaque types or value classes
+to add extra type safety to a "stringly-typed" API, by replacing
+instances of the `String` type with a more specific type.
+
+For example:
+
+```scala
+package object pkg {
+
+  import Character.{isAlphabetic, isDigit}
+
+  class Alphabetic private[pkg] (val value: String) extends AnyVal
+
+  object Alphabetic {
+    def fromString(s: String): Option[Alphabetic] =
+      if (s.forall(isAlphabetic(_))) Some(new Alphabetic(s))
+      else None
+  }
+
+  opaque type Digits = String
+
+  object Digits {
+    def fromString(s: String): Option[Digits] =
+      if (s.forall(isDigit(_))) Some(s)
+      else None
+
+    def asString(d: Digits): String = d
+  }
+}
+```
+
+The code here is relatively comparable. However, when changing
+`String` to `Alphabetic` in code, the following types would be changed
+(i.e. boxed):
+
+ * `Array[Alphabetic]`
+ * `Option[Alphabetic]` (e.g. the result of `Alphabetic.fromString`)
+ * `Vector[Alphabetic]`
+ * `Alphabetic => Boolean`
+ * `Map[Alphabetic, Int]`
+ * `Ordering[Alphabetic]`
+ * `(Alphabetic, String)`
+ * `Monoid[Alphabetic]`
+
+In many cases users won't mind the fact that this code will box, but
+there will certainly be an impact on the bytecode produced (and
+possibly the runtime performance).
+
+By contrast, replacing `String` with `Digits` is guaranteed to have no
+impact (all occurances of `Digits` are guaranteed to be erased to
+`String`). Aside from the ergonomics of calling the `fromString` and
+`asString` methods, there's no runtime impact of any kind.
```

Comment:
You should mention that this is only true when the underlying type is an `AnyRef` type. If the underlying type is a primitive, e.g., `Double`, you will trading boxing a primitive `double` inside an `Alphabetic` for boxing the same primitive `double` inside a `java.lang.Double`.

Reply:
Added a section to hopefully make this clearer.